ws-server/src/server.rs

```diff
@@ -296,34 +296,46 @@ async fn background_task(
 				}
 			}
 			Some(b'[') => {
-				if let Ok(batch) = serde_json::from_slice::<Vec<Request>>(&data) {
-					if !batch.is_empty() {
-						// Batch responses must be sent back as a single message so we read the results from each
-						// request in the batch and read the results off of a new channel, `rx_batch`, and then send the
-						// complete batch response back to the client over `tx`.
-						let (tx_batch, mut rx_batch) = mpsc::unbounded::<String>();
-
-						for fut in batch
-							.into_iter()
-							.filter_map(|req| methods.execute_with_resources(&tx_batch, req, conn_id, &resources))
-						{
-							method_executors.add(fut);
-						}
-
-						// Closes the receiving half of a channel without dropping it. This prevents any further
-						// messages from being sent on the channel.
-						rx_batch.close();
-						let results = collect_batch_response(rx_batch).await;
-						if let Err(err) = tx.unbounded_send(results) {
-							tracing::error!("Error sending batch response to the client: {:?}", err)
+				// NOTE(niklasad1): we must move deserialization into the async block below
+				// because `Request` uses `serde(borrow)` and that buffer is used again
+				// for the next iteration of the loop.
+
+				let d = std::mem::take(&mut data);
+				let r = &resources;
+				let m = &methods;
+				let tx2 = tx.clone();
+
+				let fut = async move {
+					// Batch responses must be sent back as a single message so we read the results from each
+					// request in the batch and read the results off of a new channel, `rx_batch`, and then send the
+					// complete batch response back to the client over `tx`.
+					let (tx_batch, mut rx_batch) = mpsc::unbounded();
+					if let Ok(batch) = serde_json::from_slice::<Vec<Request>>(&d) {
+						if !batch.is_empty() {
+							let futs = batch
+								.into_iter()
+								.filter_map(|req| m.execute_with_resources(&tx_batch, req, conn_id, r));
```

Comment:
Does this mean that responses could come back in a different order to the request order in the batch?

eg if we have a batch request to first `slow_hello` and then `say_hello`, and then collect those up via a channel, we'll get the `say_hello` response back first.

Do we care that the response order might not match the request order (if I'm even right about this!)?

Reply:
good question

>The Response objects being returned from a batch call MAY be returned in any order within the Array. The Client SHOULD match contexts between the set of Request objects and the resulting set of Response objects based on the id member within each Object.

I should have put a comment in the code about this but it's kind of an optimization :)

Reply:
Ah awesome, gotcha!